larpandoracontent/LArHelpers/LArPfoHelper.cc

```diff
@@ -472,10 +472,20 @@ const Vertex *LArPfoHelper::GetVertex(const ParticleFlowObject *const pPfo)
     if (pPfo->GetVertexList().empty())
         throw StatusCodeException(STATUS_CODE_NOT_FOUND);
 
-    if (pPfo->GetVertexList().size() != 1)
-        throw StatusCodeException(STATUS_CODE_FAILURE);
+    const Vertex *pVertex(nullptr);
 
-    const Vertex *const pVertex = *(pPfo->GetVertexList().begin());
+    // ATTN : Test beam parent pfos contain an interaction and start vertex
+    if (LArPfoHelper::IsTestBeam(pPfo) && pPfo->GetParentPfoList().empty())
+    {
+        pVertex = LArPfoHelper::GetVertexType(pPfo->GetVertexList(), VERTEX_START);
+    }
+    else
+    {
+        if (pPfo->GetVertexList().size() != 1)
+            throw StatusCodeException(STATUS_CODE_FAILURE);
+
+        pVertex = *(pPfo->GetVertexList().begin());
+    }
```

Comment:
We could make this cleaner if we risk some restructuring (not clear whether we should do this):

If the parent pfo list is empty, could call the same function, but with different labels: VERTEX_START for test-beam, and VERTEX_INTERACTION for other cases.

If the parent pfo list isn't empty, could still just ask for the first vertex, or know that it's label should be VERTEX_INTERACTION again (I actually think these particles should now have vertices with label VERTEX_START).

Would we still want to enforce the size of vertex list being 1 for everything except test beam particles?

Is this all too fiddly?

Reply:
I think you're right.  I didn't particularly like this bit of the function as it is ugly, but it was necessary because we've been assuming all PFOs have a single vertex everywhere else in the reconstruction.  I also agree that VERTEX_START makes more sense too for everything apart form the neutrino/test beam parent.  

From a pragmatic sense I'm tempted to leave things as they are for now to avoid the fiddly-ness and have a review of the use of vertices.  Long term this will be needed when Pandora tried to do the reconstruct about secondary interaction vertices anyway.  With that in mind I wouldn't enforce a fixed vertex list size because the start, end and interaction are all distinct valid vertices that might be of interest.  If you think it's worth doing that sooner rather than later I'm happy to give the review a go, but it might be sensible to make that a separate development from this also.  What do you think it best to do going forward with this one?